### ISSUE TITLE
read.adp.rdi() should be able to parse large files

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1093]
I have a file (coincidentally one of those new "SentinelV" files, a la #1058) which is 5.6 GB on disk. Not surprisingly, when I try and read it, I get:

``` r
> d <- read.oce('NLIW_30m 20160613T140449.pd0', from=1, to=1000, debug=10)
read.oce("NLIW_30m 20160613T140449.pd0", ...) inferred type="adp/rdi"
read.adp.rdi(...,from= 1 ,to= 1000 ...) {
  file.size= 5851155138 
Error: cannot allocate vector of size 5.4 Gb
```

The issue of course is that we just drink in the entire file as a buffer to parse, even if only a subset is desired. This is done because otherwise there is no way to know how to subset the profiles which are actually desired (either by profile number or by time).

It makes sense to have the reading be a little smarter, to at least try and parse the approximate location in the file from the info in the header. For example, the header gives the number of bytes (B) per ensemble, so if we know that and the profile number (N) that is desired we should be able to just read a chunk out of the middle of the file by guessing the location based on NxB.

**Comment by dankelley** [comment]
Hm. Can you do a `traceback` so I can see the line number where the alloc is done? My guess is that it's in the `readBin` that immediately follows that debugging line.  That means we'd need to have some pretty major changes. Note that C and read through a file by itself, so the code in `src/ldc_rdi.c` could be made to work with this. Then, I guess we could only do the `readBin` from the start of the file to the profile just after the `to` profile. This is going to be tricky if they specify `from` and `to` by times, though, because it would require decoding the times in the C code, which might be messy.

PS. There is a *lot* of expansion of memory usage in creating the R objects (e.g. things are stored in double-precision, or 8 bytes) so this file of yours would likely require a pretty large machine to read fully.

**Comment by richardsc** [comment]
Not sure what's different, but I ran the above code again inside of RStudtio (instead of a terminal R), and got the following:
```
d <- read.oce('~/research/rbrThermalMass/wireWalker/data/ADCP/NLIW_30m 20160613T140449.pd0', from=1, to=1000, debug=10)
read.oce("~/research/rbrThermalMass/wireWalker/data/ADCP/NLIW_30m 20160613T140449.pd0", ...) inferred type="adp/rdi"
read.adp.rdi(...,from= 1 ,to= 1000 ...) {
  file.size= 5851155138 
decodeHeaderRDI() {
  bytesPerEnsemble= 2400 
  numberOfDataTypes= 15 
  haveActualData= TRUE 
  dataOffset= 36 96 162 916 1294 1672 1712 1902 1998 2094 2110 2168 2184 2218 2222 
Detected dataType 0x00 0x70 for Sentinel V series configuration
  Fixed Leader Data: 00 00 2f 11 cb 57 00 35 04 5e 01 00 32 00 64 00 01 00 02 00 00 00 00 00 00 00 00 00 00 00 7d 7d a8 00 39 00 00 00 ff 00 1e 00 30 31 66 34 31 64 31 34 00 00 ff 00 81 5f 00 00 19 00 
  firmwareVersion= 47.17 (numerically, it is 47.17 )
  FLD[4]= 00010001 (looking near the systemConfiguration bytes to find a problem)
  FLD[5]= 11001011 (should be one of the systemConfiguration bytes)
  FLD[6]= 01010111 (should be one of the systemConfiguration bytes)
  FLD[7]= 00000000 (looking near the systemConfiguration bytes to find a problem)
  bits: 011 so frequency= 500 
  systemConfiguration: 11001011-01010111 
  bits: 11 Expect 111 for 25 degrees
  bits= 111 so beamAngle= 25 
  bits= 1 so beamPattern= convex 
  bits= 1 so that orientation= upward 
  numberOfBeams 4 
  numberOfCells 94 
  headingAlignment= 0 ; headingBias= 0 
  CPU.BOARD.SERIAL.NUMBER = '484910252491004952'
  SERIAL NUMBER 24449 from bytes ( 81 5f 00 00 )
  FLDLength 60  (expect 59 for Workhorse, or 50 for Surveyor/Observer)
  Variable Leader Data ( 65 bytes): 80 00 01 00 10 06 0d 0e 04 36 00 00 00 00 ec 05 00 00 ff 26 7d ff 76 ff 22 00 27 07 00 00 00 00 00 00 00 c0 00 00 00 00 00 00 00 00 00 00 07 e1 01 00 00 00 00 00 00 00 00 14 10 06 0d 0e 04 36 00 
  profile time= 2016-06-13 14:04:54 (year= 2016 month= 6 day- 13 hour= 14 minute= 4 second= 54 hundreds= 0 )
  soundSpeed=  1516 
  transducerDepth =  0 
  about to create the list to be returned
} # decodeHeaderRDI()
  about to call ldc_rdi
Error in read.adp.rdi(file, processingLog = processingLog, ...) : 
  long vectors not supported yet: memory.c:3388
```
With a traceback:
```
> traceback()
3: .Call("ldc_rdi", buf, 0) at adp.rdi.R#523
2: read.adp.rdi(file, processingLog = processingLog, ...) at oce.R#1551
1: read.oce("~/research/rbrThermalMass/wireWalker/data/ADCP/NLIW_30m 20160613T140449.pd0", 
       from = 1, to = 1000, debug = 10)
```

Just for fun, I used the linux `split` command to break the original file into 1GB chunks, and I am able to read the first of those no problem. I can only read the first, because of course it doesn't split on the 7f7f  code, so the other files have ragged starts that don't obey the rules for an RDI ADCP file ... I'm trying to figure out now how I can split the file based on those codes ...

**Comment by richardsc** [comment]
Just for documentation sake, here's the method I just used to split the file into chunks with 200000 profiles (or less).

1. I stuck a `browser()` statement at line 514 in `adp.rdi.R` so it would stop after reading the entire raw buffer (which it apparently *just* had enough memory to do) and after running `decodeHeaderRDI()`.

2. From the header, I got the `bytesPerEnsemble` field, which corresponds to the first ensemble (remember that for a sentinel file the first one is bigger than all the others). For this file that was 2400.

3. I could then decode the number of bytes per ensemble for all the other ensembles by looking at `buf[2400+2+2]` and `buf[2400+2+3]`. The `+2` is to account for the checksum at the end of the ensemble, and the `+2/+3` are the byte offsets for the `bytesPerEnsemble` field. The number of bytes per ensemble for the other ensembles is 2222 (so an offset of 2224).

4. The number of profiles in the file can be calculated from `filesize = 2402 + N * 2224`, which for this file was 2630914.

5. That means to split the first 200000 profiles off I need to split by 2402 + 2224 * 200000 = 444802402 bytes:
```
split -b 444802402 filename.pd0
```
giving a bunch of files, `xaa`, `xab`, etc. I then renamed the first one, e.g. `mv xaa datafile_001.pd0` and checked that it was a "valid" ADCP file. Check.

6. Now that the first 200000 are split off, I cat the rest of the rest by first `cat`ing the other splits together:
```
cat xa* > tmp
```
and then splitting based on the ensemble length of 2224 bytes, i.e. 2224*200000 = 444800000 bytes:
```
split -b 444800000 tmp
```

Success!

**Comment by richardsc** [comment]
Or, I could have waited to hear back from the RDI field service engineer, who pointed out that the `BBSub` program included in RDItools (free download, but requires a login) can do this too. Oh well. My way was more fun.

(On a related note, the SentinelV software "Velocity" can also do the splitting -- as well as some other processing -- but they only allow 3 licenses for each unit purchased. Oh geeze .... )

**Comment by dankelley** [comment]
So, with `BBSub`, does the present issue become of lower priority?  I might imagine that some users would need to split a 5.6Gb file because, without decimation, that 5.6Gb could expand greatly when put into general software applications (e.g. 2-byte data in the file become 8-byte data in the computer, so the 6Gb becomes more than 20Gb if it is stored in memory in floating-point form)

**Comment by richardsc** [comment]
Yes, I think this can be treated as lower priority. I haven't tested it yet (it is Windows only, sadly, as with all of RDI's software), but I was told that it should work for a sentinel file.

Interestingly, I think this has never come up before because the amount of storage in the previous version of the hardware was usually the limiting factor. E.g. it was not uncommon to have only a few hundred MB available on an SD card for storing data, so deployment planning required a careful balancing of duration, sampling frequency, and space. In space limited applications the most common approach was to do ensemble averaging *on board* the instrument. It appears that the Sentinel V units don't do the onboard averaging (I could be wrong), so they just have to have enough storage so it can be done after the fact (using non-free software ... :frowning: )

**Comment by dankelley** [comment]
A heads-up: in `dk` and `dk-1093` I am gradually working towards a `src/ldc_rdi_2.c` solution that determines index pointers to the starts of ensembles. It should be practically as fast as the old `src/ldc_rdi.c` solution (i.e. both much faster than reading the data) but it will not require constructing a buffer so it will be easier on storage, solving this bug if the extracted data will not exceed memory requirements.  

In my tests, it's giving the same values as the old code **except** that the old code erroneously has an element at the end of the index-pointer vector that equals 1. This error in finding the index-pointer probably could never matter in the real world because I think the calling R code would be immune to that kind of error. Therefore I have no plans to fix the old `src/ldc_rdi.c` code, but instead will substitute in my new code when I've tested it a bit more.

I plan to make a formal test suite (one driven by typing `make`) from our dropbox files (like for SBE) because I want to have a few tests before merging from `dk-1093` into `dk` and thereafter into `develop`.

**Comment by dankelley** [comment]
Actually, that bad end index is not always 1.  I don't know what is going on but I don't think I care, because the new code seems more sensible.  I'll be quiet on this issue now, until I post a resolution, likely tomorrow.

**Comment by dankelley** [comment]
The test suite (see #1096) now checks the adp files we have in our shared Dropbox. This is a very crude check -- I just see if I can read the files, and then I diagnose whether my new ensemble-detection code (in `src/ldc_rdi_2.c`) produces the same results as the old code (in `src/ldc_rdi.c`). I have had to make some changes to the latter, since I found some errors. As of this moment (with code not yet pushed to GH) the old and new codes agree with all but two files. And in those two files, it is pretty clear (from the uniformity of ensemble length) that the new code is working properly and that the old code is faulty.

I've looked at the code for a while now, and am not seeing why the results differ. However, at some point this becomes a loser's game: I would *like* to find out why the old code is faulty, just because I'm stubborn, but there is little practical reason in wasting time on code that I am growing certain is going to be replaced.

Why do I say replaced? Because the new code passes through the file without creating much storage, and so it will be the first step in making progress on the present issue. Besides, the new code is written as a single-pass scheme, so it is a lot more elegant.

The only possible disadvantage of the new code is that it is doing memory allocation, but I am not worried about that because I'm doing it through R wrappers (which should handle errors) and because it's been tested a fair bit (I made tiny original buffers, and monitored as the scheme automatically grew those buffers).

**tldnr** sometime today, I will merge this new ensemble-detection code into `develop`. I don't want to wait too long because both CR and I may be working on parts of the related `adp` R code, and we don't want conflicts.

PS. I cleaned up a few bugs in the `R/adp.rdi.R` code, e.g. a failure to properly initialize the `timeStart` buffer, a typo with `object$data` instead of `object@data`, etc. And this is the kind of thing that might have CR and I working on similar bits of code, so I have that in mind for my plan to merge today.

**Comment by dankelley** [comment]
One more thing, the below illustrates that reading the entirety of these long files will be a problem, because (I think) R uses the C type `int` (not `long int`) for integers. I know there is a move to go to `long int`, and that is being done gradually in various parts of R, but I don't know whether it is possible at the present time (on *any* machine) to allocate a `raw` vector of length exceeding 4.3Gb. This is not a limitation of the memory on the machine, but rather on the maximum value of an integer in R.
```R
> (2^32-1)/1e9
[1] 4.294967
```

**Comment by dankelley** [comment]
Clark, you might want to pull the `dk-1093` branch, and then do as below (maybe with extra RDI files if you have some). I suggest this because I have found errors in the old ensemble-finding code, and they are fixed in this branch. The branch also does tests of the new code for finding those ensemble starts ... and it is my goal to put that new code into `develop` soonish, while it is still in my head. I just cannot see any disadvantage to it, and the single-pass nature of the coding should make it easier to understand and thus less prone to bugs.  (The addition of this test suite helped me a lot -- I found several problems in the old code whilst trying to make it pass this test suite.)
```R
Emit:oce/tests_private[dk-1093]$ make adp_rdi_01.out
make: `adp_rdi_01.out' is up to date.
```

**Comment by dankelley** [comment]
@richardsc can you send me your sample file?  (I know you likely did that before, but I cannot find it.) That way I can do some trials.

The first task is to find a way to predict the problem without R crashing. Simply wrapping the buffer-allocation call in a `try` won't help much, because we are (presently!) storing the data in R's numeric format, which means 8 bytes for velocity data that are only 2 bytes in the RDI file. In other words, we will get failure for files only a quarter of the size that will be signalled by catching this buffer allocation problem. I just need to figure out how to ask R how much space it will be safe to request.

At least for a start, it will be easiest to read in a contiguous chunk of data. I think that can be done with minimal code disruption, because at the point where we read the whole file, we will instead `seek` to a certain spot (determined by the `from` argument), and then read a certain number of bytes (determined by the `to` argument). Once everything works with that scheme, we could return and permit the use of the `by` argument to gallop through the file.  **Note:** I think I'd prefer to use the `by` argument in my work, but it will be tricker to program, so I think it should come second, after a proof-of-concept with a contiguous subsection.

@richardsc do you mind if I move the oce-issue subdirectories with names starting with `10` to a new subdir called `10xx`? I am finding file listings to be very tedious and would like to organize all the subdirs into subdirs like the `11xx` that I made a week or two ago, so the basic listing will have at most 100 files until we get 10X the number of issues we have now.  If you agree, I will delete the present paragraph, plus your comment, since the present issue comments already span a lot of pages.

**Comment by dankelley** [comment]
I started a new branch called `dk-1093-large-rdi`, which "hangs off" the `dk` branch instead of hanging off the `develop` branch. (I also brought `dk-1093` up to date with `develop`, just to check against merge conflicts, but I intend to remove `dk-1093` and from now on to use branch names that have a bit of info in the title and not just a number.)

**Comment by dankelley** [comment]
Clark, maybe you can have a look at `dk`. I don't like to have too much evolution in branches, and `adp` is a pretty active topic, so I'd like to merge `dk` into `develop` within a few days. If you can do some tests of `dk`, then I'd be happier merging it into `develop`, after which this could get closed, if you please.

I think I have this working well on `dk` commit 9299b0c8e8377297c62afa6f6121face7c556302. It has involved a fairly significant rewrite of the code, and so I have not merged it into `develop` yet. However, I have run all the build checks, and the tests in `tests_private`, so I have a good confidence that I've not broken anything.  (I also checked the `data(adp)` numbers with those from the old code, and everything checks.)

The defaults are now different. If you do `read.oce("someAdpFile")`, then `read.adp.rdi()` will check the file length, and if it exceeds 1Mb then a guess will be made of ensemble length and `by` will be set up to give roughly 500 profiles after subsampling. (I say "roughly" because, for code-flow reasons, I make the calculation of `by` before the first ensemble has been scanned, so at that stage I can only guess on typical ensemble length.)

As for how this works in practice, below is what we get for a 6GB file. This time seems acceptable to me, since this is a pretty big file. Once a `plot` has been done of such a file, the user will likely want to look in more detail, and things get faster then, because the time is linear in the length of the file that is examined, and (in this new code) we do not examine the file past the `to` spot.
```R
> library(oce)
> system.time(d <- read.oce("~/sentinelADCP/NLIW.pd0"))
   user  system elapsed 
 10.802   3.097  16.496 
```

**Comment by richardsc** [comment]
Just getting around to testing this now. I haven't tried it extensively, but I have tried the original large (e.g. 6GB file) and a few other regular files and it seems to work.

I am concerned that the 1Mb limit will be too small -- realistically we can already read files that are several hundred MB without issue. The problem then is determining what the limit *should* be. Should it be dynamically calculated somehow for the system that is running the code (e.g. memory available, R resources, etc), or set at a reasonable, but not too large, limit that will catch most cases.

In any case, I think that when the `read.adp.rdi()` catches one of these "large file" cases and does the decimating, it should definitely warn about it. Otherwise one might think that they successfully read the file but only got something like 1/500th of the data!

**Comment by dankelley** [comment]
(Q for other developers at end.)

I've done some reading, and there doesn't seem to be a way for R code to determine how much memory is available. (I was hoping that `memory.limit` would help but it is only on mswindows.)

I have seen that normal 32-bit Windows has an "address-space limit" of 2Gb (see https://stat.ethz.ch/R-manual/R-devel/library/base/html/Memory-limits.html for the R docs, and perhaps see https://stat.ethz.ch/R-manual/R-devel/library/base/html/Memory-limits.html for details) and so that might be a way to set up a criterion. Since velo in the `.rda` file is 2 bytes per datum and 8 bytes when made into R "numeric", it will be the driver in memory usage. (The oce object holds e.g. backscatter as 1-byte.) I could do a calculation assuming that both `a` and `q` matrices are there. Let N be the total number of data (i.e. # cells times # times time # beams). Then the storage will be 10N bytes (8 bytes per velo datum, 1 byte for `a` and1 byte for `q`). So that sets a limit of N<2e8. Maybe call that 1e8 to be conservative, or perhaps 5e7 for a factor of 4 safety. So, that's a 50Mb file. Larger than that, and we might expect problems on a 32 bit windows machine. I know we have used larger files in osx (I think, 64 bit osx) but we are talking here about a for a warning and decimation that can always be overcome if the user specifies `from`, `by` and `to`, so I think it's ok.

From the above, I propose to set the limit to 50Mb. Files smaller than that, and we just read all the data with the default `read.oce(filename)`. Larger, and we decimate as appropriate.

OK?

**Comment by richardsc** [comment]
50 MB sounds reasonable to me, provided we spit a warning.

**Comment by dankelley** [comment]
I'll look at this today. My calculation was wrong. A particular datum in the input file is 2 bytes for velo, 1 for quality, and 1 for backscatter. In the final object, these turn into 4bytes, 1 byte, and 1 byte. Thus the input is 4 bytes, and the output is 10 bytes. There is also some expansion of other things (like time, etc) but these elements will be small compared to the matrices so I will ignore them. Still, I'll round the 10/4 up to 12/4, or 3.  Thus, where I had a 10 in the above calculation, I now think there should be a 3. That's what I'll be using. My plan is to explain the calculations in some detail in the docs (rather than putting them here, where they won't get seen by users) and then to code after that. Likely this will be today sometime.

**Comment by richardsc** [comment]
So is the file size limit going to go up, or down? Honestly, I've never had a problem reading 100+MB files with the current code. I don't want the file size limit to be so small that it's going to decimate just about every file.

I recall we have a shared directory of RDI files for testing. I'm just headed out the door, but what are the file sizes like on those?

**Comment by dankelley** [comment]
going up a factor of 10/4.  I’m actually writing the docs right now, while waiting for the first of two phone meetings this morning

**Comment by dankelley** [comment]
I have pushed to the `dk` branch commit 8cc60a7fd69d7ef508e98b5b726c929f5b0bb27d, which has an updated scheme. The help page explains the reasoning and the procedure. The relevant code starts at line 638 of `R/adp.rdi.R`. Clark, it would help a lot if you could build this version, read the help page to see that it makes sense (this is really important ... I want users to understand things based on the resource that they will find, which is that help page) and look at the code to see if my coding looks ok. At the moment, it spits out `message()` output, but I will change that to `oceDebug()` after things seem OK.

I tried this on the large file set that we have been working on. With a concatenated 5.4G file, it uses `by=5` (i.e. takes every 5th profile), yielding a 526181x94x4 array for `v`.  For the smaller 424M files, it uses `by=1` and gets a 199999x94x4 array for `v`. The ratio of the number of profiles is thus 0.38, which is of order 1, as hoped for. (I say "hoped for" because I am using a guess at bytes per profile instead of using the known number of bytes ... I can change that, of course.)

Summary -- asking for Clark to have a look at these changes, and to assess the `dk` functioning for RDI files generally. It would be great to get an idea of where we stand before the middle of next week, because the code is "in my head" now and won't be, 10 days from now.  Also, I don't like having a complex branch evolving in parallel to `develop`, if I can avoid it.

**Comment by dankelley** [comment]
Clark -- there's a question for you at the end.

I switched to 500MB and rewrote the docs a bit. Now, I get as below, for a 5.4GB file. Things are looking reasonably good to me. I note that the object ends up as about 1GB, which is about double the size that I had expected. Still, I bet most people have machines with 4GB or higher RAM, so a 1GB object in R is quite likely to be okay.
```R
> library(oce)
> system.time(d<-read.oce("~/sentinelADCP/NLIW.pd0"))
   user  system elapsed 
165.571   8.711 178.313 
Warning message:
In read.adp.rdi(file, processingLog = processingLog, ...) :
  setting by=11 for a large RDI file
> format(object.size(d), units="GB")
[1] "1.1 Gb"
```

Clark -- does this seem good to you?

**Comment by dankelley** [comment]
PS. I don't like waiting 3 minutes to read an object. I'm thinking of causing `monitor` to switch to `TRUE` for large files. Seem OK, Clark?  Also, I think I'd like to make the monitoring use conventional R progress bars, probably with `txtProgressBar()`, since users are familiar with that.  This is just a personal preference, though, and I'd like advice on it. Other oce things stay quiet, but they usually are faster, I think.

**Comment by dankelley** [comment]
I went ahead and did the progress-bar stuff. It can be undone, although I think it's quite nice to have. I think there's something weird, where the bar occurs twice, but I see that in other applications so I'm not sure it is oce-related.

**Comment by richardsc** [comment]
I like it! I don't think I'd ever used the `monitor=TRUE` before. One note -- the docs for `read.adp.rdi()` need to be updated to reflect the new `monitor` behaviour (probably should be made more general).

**Comment by richardsc** [comment]
I think this can be closed now, as it is addressed as stated. What do you think @dankelley ?

**Comment by dankelley** [comment]
I agree to close ... but just want to check with Clark -- do you agree that I now merge `dk` into `develop`?

**Comment by richardsc** [comment]
Yup, merge it. We can deal with that other winriver issue ( #1168) from develop (or another branch) if we need to.

**Comment by dankelley** [comment]
I merged `dk` into `develop` and pushed to GH. The result passes (a) Rstudio "check", (b) Rstudio "test" and (c) local tests in `tests_private`.  Of these, I'd say (c) are the most relevant because it checks 19 different RDI files. I suspect it will go through the travisCI checks within the hour.

I'm going to close this, to save Clark the hassle (since he already agreed it should be closed).
